docs/rfc/0000-cli-infrastructure.adoc

```diff
@@ -0,0 +1,254 @@
+functionality of `rad` subcommands. The majority of subcommands that
+interact with the storage can be done while the daemon is offline. The
+only action the author can currently think of — that is online — is
+replication of a URN from the network.
+
+This leaves us in a predicament of thinking about how multiple Radicle
+applictions may interact with each other. For example, if Upstream is
+running and I create a project via the CLI, then a good user
+experience is that the Upstream is updated with the new project
+(without refreshing). In RFC-682, it is alluded that a PubSub
+component could solve this, however, that requires the daemon to be
+online. The author prefers not to specify the PubSub component
+here. To bridge the gap, we recommend that a subcommand can add a
+parameter `daemon-mode := offline (default) | online`. If set to
+`online` the command will see if the daemon is online, and if not ask
+it to come online before applying the command.
+
+== Key Access
+
+RFC-682#Signatures outlined the signing of Radicle data. This is essential to
+this RFC as commands that write to the storage generally need access
+to the secret key, whether for signing data or proof of ownership to
+the storage. For most (if not all) read-only commands, no access to
+the secret key is needed.
+
+To achieve the goals of having a read-only storage and using
+`<<ssh-agent>>` for signing, we propose the following should be
+implemented:
+
+. `ssh-agent` compatability with `<<radicle-keystore>>`, which has been
+started in `<<rk-17, radicle-keystore#17>>`.
+. A subcommand `rad-key-ssh-add` for adding one's secret key to the
+`ssh-agent` using the supplied constraints, e.g. ask for password.
+. `radicle-link#461` to have a read-only storage.
+
+== Library/Command Organisation
+
+As proposed here, `radicle-link` will provide a core set of
+functionality. Something this document wishes to standardise is the
+organisation of these components. The author sees two options:
+. One binary that defines the core subcommands
+. Many binary that take advantage of the subcommand mechanism
+
+Pros of *1.*:
+
+* One-stop shop for all core functionality
+* One binary is packages and distributed
+
+Cons of *1.*:
+
+* Making changes to one subset of the commands could mean a version
+  change of the whole binary
+
+Pros of *2.*:
+
+* We will be organising commands as libraries (as per RFC-682) — so
+  modular commands would mirror this structure well.
+* Subcommands can be versioned independently of each other.
+* We can dogfood our subcommand functionality
+
+Cons of *2.*:
+
+* Makes it easier to override one of the core commands.
+* Many binary must be packaged and distributed.
+
+The author recommends *2.*. The modularity provides a clean separation
+of concerns and avoids a monolothic approach. We assume good faith
+of overrides not happening [TODO: is there some way we can just
+disallow this programatically?]. For packaging and distribution of the
```

Comment:
Instead, you may detail how development can be made more convenient. Eg. make it so development builds always produce binaries with a `-dev` suffix, and provide `systemd`/`launchd` definitions so they either call `cargo build` or reference the `-dev` binaries.

Reply:
Ah so, there would be the `rad` binary, but then there would also be a `rad-key-dev` binary, for example?

Reply:
Thinking about it, maybe the RFC is not the right place to talk about this (the same thing could also be achieved using `nix` and / or `direnv`). Maybe just assume proper packaging.